### PR TITLE
Authentication

### PR DESCRIPTION
**Pull request by Abomy**

### CODE REVIEW

File: foundationz_backend/src/modules/auth/auth.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+const { Strategy, ExtractJwt, } = passportJWT
+
+const params = {
+    secretOrKey: process.env.JWT_SECRET,
+    jwtFromRequest: ExtractJwt.fromAuthHeaderAsBearerToken()
+}
+
+export const authStrategy = new Strategy(params, (payload, done) => {
+    console.log("test")
```
Comment (Kysumi) [inline foundationz_backend/src/modules/auth/auth.ts:13]: console log

File: foundationz_backend/src/modules/auth/auth.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import passportJWT from 'passport-jwt'
+import {User} from "@orm/user";
+import jwt from "jsonwebtoken";
+
+const { Strategy, ExtractJwt, } = passportJWT
+
+const params = {
```
Comment (Kysumi) [inline foundationz_backend/src/modules/auth/auth.ts:7]: nitpick: could be inline.

File: foundationz_backend/package.json [hunk 3]
```diff
@@ -8,7 +8,8 @@
   "scripts": {
     "test": "echo \"Error: no test specified\" && exit 1",
     "old": "nodemon --watch './**/*.{ts,graphql}' --exec \"node --require dotenv/config --experimental-specifier-resolution=node --loader ./loader.js \" src/index.ts",
-    "dev": "tsc-watch --onSuccess \"npm run watch\"",
+    "olddev": "tsc-watch --onSuccess \"npm run watch\"",
+    "dev": "concurrently \"tsc-watch\"  \"npm run watch\"",
```
Comment (Kysumi) [inline foundationz_backend/package.json:12]: legend ty

File: foundationz_backend/src/modules/auth/crypto.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
+import { User } from "@orm/user";
```
Comment (Kysumi) [inline foundationz_backend/src/modules/auth/crypto.ts:1]: files formatting looks bit off

File: foundationz_backend/src/modules/graphql/auth/auth.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
+import {extendType, nonNull, objectType, stringArg} from "nexus";
+import {User} from "@orm/user";
+import {ApolloError} from "apollo-server-core";
+import {validatePassword} from "@auth/crypto";
+import {signToken} from "@auth/auth";
+
+export const UserLogin = extendType({
+    type: "Query",
```
Comment (Kysumi) [inline foundationz_backend/src/modules/graphql/auth/auth.ts:8]: not sure of best practices for logging in on a GQL API. Kinda would have expected a mutation but I guess a query also kind of makes sense?

Is this something you found?

File: foundationz_backend/src/modules/graphql/auth/auth.ts [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+                if(validatePassword(user,password)){
+                    return [{
+                        id: user.id,
+                        email: user.email,
+                        token: signToken(user)
+                    }]
+                }
+                throw new ApolloError(`Invalid email or password!` );
+
```
Comment (Kysumi) [inline foundationz_backend/src/modules/graphql/auth/auth.ts:31]: empty lines

File: foundationz_backend/src/modules/graphql/auth/auth.ts [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+                }
+
+                if(validatePassword(user,password)){
+                    return [{
+                        id: user.id,
+                        email: user.email,
+                        token: signToken(user)
+                    }]
+                }
+                throw new ApolloError(`Invalid email or password!` );
```
Comment (Kysumi) [inline foundationz_backend/src/modules/graphql/auth/auth.ts:30]: We should look into a standardised way for handling/structuring server-side validation

https://foundationz.youtrack.cloud/issue/CW-8

File: foundationz_backend/src/modules/graphql/auth/auth.ts [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+            type: "UserAuth",
+            args: {
+                email: nonNull(stringArg()),
+                password: nonNull(stringArg()),
+            },
+            resolve: async (_, { email, password }) => {
+                const user = await User.query().where('email', email).first();
+
+                if(!user){
+                    throw new ApolloError(`User with the email ${email} not found!` )
```
Comment (Kysumi) [inline foundationz_backend/src/modules/graphql/auth/auth.ts:20]: ` if (!user) throw new AuthenticationError('you must be logged in');`

https://www.apollographql.com/docs/apollo-server/security/authentication/

Looks like ApolloServer has a error type for this

File: foundationz_backend/src/modules/graphql/auth/auth.ts (hunk 7 above)
Comment (Abomy) [reply foundationz_backend/src/modules/graphql/auth/auth.ts:30]: I second that

**Pull request merged**